### PR TITLE
UDP Tunneling: Optionally propagate response headers and trailers to downstream info

### PR DESCRIPTION
**Pull request by IssaAbuKalbein**
Commit Message: Optionally propagate response headers and trailers to downstream info
Additional Description: Add support for saving upstream response headers and trailers to downstream info
Risk Level: low
Testing: integration tests
Docs Changes: API
Release Notes: None
Platform Specific Features: None

### CODE REVIEW

**Comment by IssaAbuKalbein** [comment]
ping @botengyao

File: test/integration/udp_tunneling_integration_test.cc [hunk 1]
```diff
@@ -712,6 +720,94 @@ TEST_P(UdpTunnelingIntegrationTest, ConnectionAttemptRetry) {
...
+                    "",
+                    BufferOptions{1, 30},
+                    absl::nullopt,
+                    session_access_log_config,
+                    true,
+                    false};
+  setup(config);
+
+  const std::string datagram = "hello";
+  establishConnection(datagram);
```
Comment (botengyao) [inline test/integration/udp_tunneling_integration_test.cc:753]: Could you also one test case to cover invalid response?

**Review by botengyao — COMMENTED** [review]
Thanks for adding this support! LGTM module one test addition. 

Besides, could you also update the changelogs [here](https://github.com/envoyproxy/envoy/blob/main/changelogs/current.yaml#L92)?

/wait

File: test/integration/udp_tunneling_integration_test.cc (hunk 1 above)
Comment (IssaAbuKalbein) [reply test/integration/udp_tunneling_integration_test.cc:753]: Added test.

**Comment by botengyao** [comment]
` bazel run //tools/code_format:check_format -- fix` is pretty useful and just in case 0/

/wait

File: test/integration/udp_tunneling_integration_test.cc [hunk 2]
```diff
@@ -712,6 +720,141 @@ TEST_P(UdpTunnelingIntegrationTest, ConnectionAttemptRetry) {
...
+    typed_config:
+      '@type': type.googleapis.com/envoy.extensions.access_loggers.file.v3.FileAccessLog
+      path: {}
+      log_format:
+        text_format_source:
+          inline_string: "%FILTER_STATE(envoy.udp_proxy.propagate_response_headers:TYPED)%\n"
+)EOF",
+                                                            access_log_filename);
+
+  TestConfig config{"host.com",
```
Comment (botengyao) [inline test/integration/udp_tunneling_integration_test.cc:739]: nit: const

File: test/integration/udp_tunneling_integration_test.cc [hunk 3]
```diff
@@ -712,6 +720,141 @@ TEST_P(UdpTunnelingIntegrationTest, ConnectionAttemptRetry) {
...
+    typed_config:
+      '@type': type.googleapis.com/envoy.extensions.access_loggers.file.v3.FileAccessLog
+      path: {}
+      log_format:
+        text_format_source:
+          inline_string: "%FILTER_STATE(envoy.udp_proxy.propagate_response_trailers:TYPED)%\n"
+)EOF",
+                                                            access_log_filename);
+
+  TestConfig config{"host.com",
```
Comment (botengyao) [inline test/integration/udp_tunneling_integration_test.cc:828]: nit: const

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto [hunk 4]
```diff
@@ -160,6 +160,14 @@ message UdpProxyConfig {
     // while the upstream is not ready will be dropped. In case this field is set but the options
     // are not configured, the default values will be applied as described in the ``BufferOptions``.
     BufferOptions buffer_options = 9;
+
+    // Save the response headers to the downstream info filter state for consumption
+    // by the session filters. The filter state key is ``envoy.udp_proxy.propagate_response_headers``.
+    bool propagate_response_headers = 10;
```
Comment (htuch) [inline api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:166]: How is this intended to be used? Should docs be updated?

Comment (IssaAbuKalbein) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:166]: It's intended to be used for receiving response headers/trailers from upstream, which provides additional metadata about the upstream endpoint and connection status that can be used for telemetry purposes.
It adds similar functionality to udp_tunneling as in tcp_tunneling:
- [tcp tunneling: optionally move response header to downstream info #23118
](https://github.com/envoyproxy/envoy/pull/23118)
- [TCP Tunneling: Optionally propagate response trailers to downstream info #25023
](https://github.com/envoyproxy/envoy/pull/25023)

Comment (htuch) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:166]: Makes sense, thanks.

File: source/extensions/filters/udp/udp_proxy/config.h [hunk 5]
```diff
@@ -14,6 +14,43 @@ namespace UdpProxy {
...
+class TunnelResponseHeadersOrTrailers : public StreamInfo::FilterState::Object {
+public:
+  ProtobufTypes::MessagePtr serializeAsProto() const override;
+  virtual const Http::HeaderMap& value() const PURE;
+};
+
+/**
+ * Response headers for the tunneling connections.
+ */
+class TunnelResponseHeaders : public TunnelResponseHeadersOrTrailers {
```
Comment (htuch) [inline source/extensions/filters/udp/udp_proxy/config.h:29]: @kyessenov this is basically replicating what you did in https://github.com/envoyproxy/envoy/pull/23118 but for UDP. Looks fine to me, do you have any opinion on whether it is worth refactoring and sharing code across the implementations or just going with what is in this PR?

Comment (IssaAbuKalbein) [reply source/extensions/filters/udp/udp_proxy/config.h:29]: ping @kyessenov

Comment (kyessenov) [reply source/extensions/filters/udp/udp_proxy/config.h:29]: Sorry, was OoO. I'm fine if we postpone code consolidation later. These are tiny classes and fairly easy to refactor.

Comment (IssaAbuKalbein) [reply source/extensions/filters/udp/udp_proxy/config.h:29]: Thanks @kyessenov .
@htuch Do you have other comments on the PR, please?

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**